Mathlib/Algebra/Squarefree/UniqueFactorizationDomain.lean

```diff
@@ -0,0 +1,105 @@
+/-
+Copyright (c) 2024 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import Mathlib.Algebra.Squarefree.Basic
+import Mathlib.RingTheory.UniqueFactorizationDomain
+
+/-!
+# Square-free elements of unique factorization monoids
+
+## Main results:
+ * `squarefree_mul_iff`: `x * y` is square-free iff `x` and `y` have no common factors and are
+   themselves square-free.
+
+-/
+
+variable {R : Type*} [CancelCommMonoidWithZero R] [UniqueFactorizationMonoid R] {x y p d : R}
+
+open UniqueFactorizationMonoid
+
+theorem squarefree_iff (hx₀ : x ≠ 0) (hxu : ¬ IsUnit x) :
+    Squarefree x ↔ ∀ p, Prime p → ¬ (p * p ∣ x) := by
+  refine ⟨fun h p hp hp' ↦ hp.not_unit (h p hp'), fun h d hd ↦ ?_⟩
+  induction' d using induction_on_prime with u hu z p _ hp _
+  · obtain ⟨p, hp⟩ := exists_prime_iff.mpr ⟨x, ⟨hx₀, hxu⟩⟩
+    rw [mul_zero, zero_dvd_iff] at hd
+    simpa [hd] using h p hp
+  · assumption
+  · specialize h p hp
+    replace hd : p * p ∣ x := by rw [mul_mul_mul_comm] at hd; exact dvd_of_mul_right_dvd hd
+    contradiction
+
+namespace Squarefree
+
+theorem pow_dvd_of_squarefree_of_pow_succ_dvd_mul_right {k : ℕ}
+    (hx : Squarefree x) (hp : Prime p) (h : p ^ (k + 1) ∣ x * y) :
+    p ^ k ∣ y := by
+  nontriviality R
+  rcases eq_or_ne x 0 with rfl | hx₀
+  · have := not_squarefree_zero hx; contradiction
+  wlog hxu : ¬ IsUnit x
+  · rw [not_not] at hxu; rw [hxu.dvd_mul_left] at h; rw [pow_succ] at h; exact dvd_of_mul_left_dvd h
+  rw [squarefree_iff hx₀ hxu] at hx
+  by_cases hxp : p ∣ x
+  · obtain ⟨x', rfl⟩ := hxp
+    have hx' : ¬ p ∣ x' := by
+      specialize hx p hp
+      contrapose! hx
+      exact mul_dvd_mul_left p hx
+    replace h : p ^ k ∣ x' * y := by
+      rw [pow_succ, mul_assoc] at h
+      exact (mul_dvd_mul_iff_left hp.ne_zero).mp h
+    exact hp.pow_dvd_of_dvd_mul_left _ hx' h
+  · exact dvd_trans (pow_dvd_pow _ (k.le_add_right 1)) (hp.pow_dvd_of_dvd_mul_left _ hxp h)
```

Comment:
Can this be simplified by having a lemma saying that if `p ^ k | x * y`, then there exist `i, j` with `i + j = k` such that `p ^ i | x` and `p ^ j | y`?

Reply:
Such a lemma would indeed shorten this proof (and it is worth having) but it is a moderate amount of work.

Inspired by this remark I have however pushed a golf which cuts away more than half of this proof: https://github.com/leanprover-community/mathlib4/pull/10170/commits/27cdca5d4b42c4e383ffa2a7ab4b39de92027d55